File: fsharp/hamming/Hamming.fs
module Hamming

let private dissimilar (a: 'a) (b: 'a) =
    match a = b with
    | true -> 0
    | false -> 1

let distance (strand1: string) (strand2: string): int option =
    let bases1: char array = strand1.ToCharArray()
    let bases2: char array = strand2.ToCharArray()
    match bases1.Length = bases2.Length with 
    | true ->
        Array.fold2 (fun (distance: int) (base1: char) (base2: char) -> distance + dissimilar base1 base2) 0 bases1 bases2
        |> Some
    | false -> None

File: solutions/scala/hamming/1/src/main/scala/Hamming.scala
object Hamming
{
  private def dissimilar[A] (a: A, b: A):Int =
  a == b match
  {
    case true => 0
    case false => 1
  }

  def distance(dnaStrandOne: String, dnaStrandTwo: String): Option[Int] =
  {
    def bases1: Array[Char] = dnaStrandOne.toCharArray()
    def bases2: Array[Char] = dnaStrandTwo.toCharArray()
    bases1.length == bases2.length match
    {
      case true => Some(bases1.zip[Char](bases2).map[Int](Function.tupled(dissimilar[Char])).sum[Int])
      case false => None
    }
  }
}
